add projects for maven, gradle and sbt for java and scala

jtgrabowski opened the pull request:
assembly works for all of them
sbt and gradle can figure out DSE classpath (thanks Russ)

bcantoni commented:
Are there meant to be Gradle examples as well? This PR seems to remove them all.

jtgrabowski commented:
@bcantoni  yes, there are moved to `scala/gradle` and `java/gradle` dirs

RussellSpitzer commented:
These should all be converted to use the now publicly available DataStax Spark dependency that Artem and Predrag built

jtgrabowski commented:
will do

bcantoni commented:
In this PR I see two build.gradle files which are being deleted, but none which are being added or moved. Under scala and java I just see maven and sbt dirs.

Regarding the readme file and discussion of dependencies, would it help to have a few concrete examples with real version numbers listed? This seems to be a tricky part for people, so I think a couple of real examples here would be helpful, like:
* With DSE
  * DSE 5.0.0
  * DSE 4.8.8
  * DSE 4.7.8
* Without DSE
  * Spark 2.0, Cassandra 3.0, Connector 1.6
  * Spark x, Cassandra y, Connector z
  * etc.

jtgrabowski committed:
add projects for maven, gradle and sbt for java and scala

assembly works for all of them
sbt and gradle can figure out DSE classpath (thanks Russ)

jtgrabowski commented:
@bcantoni you are right, sorry, my .gitignore file ate gradle examples

jtgrabowski committed:
Change project structure to dse / oss

dse is for users with DSE environment (these projects use
DSE connector dependency)

oss id for open source users with manual dependencies

jtgrabowski commented:
@RussellSpitzer moved to publicly available DataStax Spark dependency that Artem and Predrag built

@bcantoni channged to project structure to dse vs oss. dse uses dependency that Russ mentioned, which is available only for 5.0.1. Do we want to have "custom" examples for older versions of DSE?
For oss the [version compatibility table](https://github.com/datastax/spark-cassandra-connector#version-compatibility) should be sufficient. We could mention it in the readme file, wdyt?

bcantoni commented:
The version compatibility table on the connector site is good and we can point to it. However it (intentionally) does not list the DSE version.

In any case if I take a step upwards, I think we want to answer the use case of "I'm using DSE version x.y, how do I set up my Spark project to work with it?". A table full of versions would probably be impossible to maintain, so maybe just an explanation of how to find and set the Spark and Spark Connector versions based on the DSE version you're using. The [DSE Release Notes](http://docs.datastax.com/en/datastax_enterprise/4.8/datastax_enterprise/RNdse.html) pages might be helpful in this regard.

RussellSpitzer commented on java/gradle/dse/build.gradle:
```diff
@@ -0,0 +1,33 @@
+    }
+}
+
+configurations {
```
May want to add a quick comment "The provided configuration behaves the same as the sbt "provided" keyword which will cause jars to be excluded from assembled fat-jar"

RussellSpitzer replied:
This example `jar` command is going to technically build a fat jar incase that wasn't clear :)

RussellSpitzer commented on python/upercasing_with_spark.py:
```diff
@@ -0,0 +1,16 @@
+
+    letters = ['d', 'a', 't', 'a', 's', 't', 'a', 'x']
+
+    rdd = sc.parallelize(letters)
```
Should do a python DF from Cassandra :)

RussellSpitzer commented on scala/sbt/dse/src/main/scala/com/datastax/spark/example/WriteRead.scala:
```diff
@@ -20,29 +18,28 @@ object Example extends App {
-
-  // This Hive Context will use the DSE HiveMetaStore
-  val sqlContext = new HiveContext(sc)
+  val sqlContext = new CassandraSQLContext(sc)
```
Nooooooooooooooooooooooooooooooooooooooooooooooooo
HiveContext Only!

RussellSpitzer replied:
Several Reasons we Don't Want Anyone using CSC

1) Limited featureset
2) No longer really supported by spark
3) HiveContext is better in every way

RussellSpitzer replied:
Also you don't actually need the metastore for the HiveContext

RussellSpitzer commented:
I left a few notes

jtgrabowski committed:
move to hivecontext and fix oss builds

jtgrabowski commented:
@RussellSpitzer  applied your remarks

@bcantoni at the moment there is no way to use current `dse` template for DSE version lower than 5.0.0 (no connector artifact published). We have two options here.
1. Restore previous project template for DSE < 5.0.0
2. Talk to Artem if it is possible to publish connector artifact for DSE < 5.0.0 and if it is, use current project template for all DSE versions

I'll talk to Artem when he is back from vacation and get back to you. Let's treat (1) as a backup plan.

jtgrabowski commented on java/sbt/dse/src/main/java/com/datastax/spark/example/WriteRead.java:
```diff
@@ -58,15 +56,19 @@ public Object apply(Session session)
-    sqlContext.setKeyspace("ks");
-    List<Tuple2<Integer, Integer>> sqlReadData = sqlContext.sql("SELECT * FROM kv").javaRDD()
+    // Read Data Using the Hive Context
+    List<Tuple2<Integer, Integer>> sqlReadData = hiveContext
```
Had to ditch select statement because I couldn't get rid of the exception "table kv.ks couldn't be found".

RussellSpitzer replied:
Without DSE the metastore won't exist, but you can make an entry by registering a DataFrame or by using the following syntax

https://github.com/datastax/spark-cassandra-connector/blob/master/doc/14_data_frames.md#creating-dataframes-using-spark-sql

RussellSpitzer commented on scala/maven/oss/pom.xml:
```diff
@@ -12,7 +12,11 @@
-  <!-- Please make sure that following dependencies have versions corresponding to the ones in your cluster. -->
+  <!--
+    Please make sure that following dependencies have versions corresponding to the ones in your cluster.
+    Note that spark-hive and spark-cassandra-connector should be provided with '&#45;&#45;packages' flag
```
spark-hive should not be necessary in --packages

RussellSpitzer commented on java/sbt/oss/build.sbt:
```diff
@@ -7,9 +7,12 @@ crossPaths := false
 libraryDependencies ++= Seq(
-  "org.apache.spark" %% "spark-core" % "1.6.1" % "provided",
-  "org.apache.spark" %% "spark-sql" % "1.6.1" % "provided",
+  "org.apache.spark" %% "spark-core" % "1.6.2" % "provided",
```
not super important but we could always have sparkVersion as a variable to avoid repeating

RussellSpitzer commented on scala/gradle/oss/build.gradle:
```diff
@@ -6,15 +6,20 @@ repositories {
 }
 
 // Please make sure that following dependencies have versions corresponding to the ones in your cluster.
+// Note that spark-hive and spark-cassandra-connector should be provided with '--packages' flag to
```
spark-hive should not be on this list

RussellSpitzer commented on scala/sbt/dse/src/main/scala/com/datastax/spark/example/WriteRead.scala:
```diff
@@ -1,10 +1,9 @@
+import org.apache.spark.{SparkConf, SparkContext}
 
 object WriteRead extends App {
 
```
May also want to add a note (just because we just got another support ticket about this) that in DSE it is not necessary to set an connection parameters for spark.master (since it will be done automatically)

RussellSpitzer commented:
I'm +1, This is a great starting point for our users. We can always do more tweaks later

jtgrabowski committed:
review remarks

jtgrabowski commented:
Russ, could you please merge this (I do not have required privileges)?
